components/availability/src/Availability.svelte

```diff
@@ -164,33 +164,50 @@
     });
   //#endregion layout
 
-  function handleTimeSlotClick(selectedSlot: any): string {
-    if (selectedSlot.selectionStatus === "unselected") {
-      if (click_action === "choose") {
-        sendTimeSlot(selectedSlot);
-      }
-      slotSelection = [...slotSelection, selectedSlot];
-      return "selected";
+  //#region event query
+  let query: Availability.EventQuery;
+  $: query = {
+    component_id: id,
+    access_token: access_token,
+    calendar_id: "",
+    participants: [{ email_address: "" }],
+  };
+  //#region event query
+
+  //#region booking event logic for single time slot and consecutive time slots
+  $: sortedSlots = [
+    ...slotSelection.sort((a, b) => a.start_time > b.start_time),
+  ].reduce((events, currentTimeSlot, i) => {
+    if (i === 0) {
+      events = [currentTimeSlot];
     } else {
-      slotSelection = slotSelection.filter(
-        (chosenSlot) => chosenSlot != selectedSlot,
-      );
-
-      return "unselected";
+      let lastMergedSlot = events[events.length - 1];
+      if (currentTimeSlot.start_time <= lastMergedSlot.end_time) {
+        lastMergedSlot.end_time = new Date(
+          Math.max(lastMergedSlot.end_time, currentTimeSlot.end_time),
+        );
+      } else {
+        events = [...events, currentTimeSlot];
+      }
     }
+    return events;
+  }, []);
+
+  function toggleSelectedTimeSlots(selectedSlot: obj): obj[] {
+    return (slotSelection =
+      selectedSlot.selectionStatus === "selected"
+        ? [...slotSelection, selectedSlot]
+        : slotSelection.filter((slot) => slot != selectedSlot));
   }
 
-  function sendTimeSlot(selectedSlot: Availability.TimeSlot) {
-    let start_time = new Date(selectedSlot.start_time);
-    let end_time = new Date(selectedSlot.end_time);
-    const timeslot: Availability.TimeSlot = {
-      start_time,
-      end_time,
-    };
-    dispatchEvent("timeSlotChosen", {
-      timeslot,
-    });
+  function sortAndSetEvent(slots: obj[]) {
+    dispatchEvent("timeSlotChosen", { timeSlots: [...slots] });
+    if (allow_booking) {
+      EventStore.createEvent(slots, query); // currently doesnt' work as calendar ID is not avaialble (To be completed by a different story)
+    }
+    slotSelection = [];
```

Comment:
One thing I noticed about setting slotSelection ere is that it doesn't de-select the slots themselves.

I'd like to see `slotSelection` act as a reactive variable -- maps/filters over days, returning any slots with a class of "selected" -- rather than a declarative array that you can push into. I think that'll reduce eventual cases where our states are out of sync (example: conditions where a server/middleware is polled and returns newly-booked timeslots that disagree with your selections)

Reply:
Made a story: https://app.clubhouse.io/nylas/story/65417/availability-component-make-slots-reactive